File: 207-course-schedule/207.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        marked = {}
        in_stack = {}

        DG = [[] for _ in range(numCourses)]
        for u,v in prerequisites:
            DG[v].append(u)

        def dfs(v):
            in_stack[v] = True
            marked[v] = True

            for w in DG[v]:
                if not w in marked:
                    if dfs(w):
                        return True
                else:
                    if w in in_stack and in_stack[w]:
                        return True

            in_stack[v] = False
            return False

        for v in range(numCourses):
            if not v in marked:
                if dfs(v):
                    return False

        return True

File: 207-course-schedule/207.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> graph(numCourses, vector<int>());
        vector<int> in_degree(numCourses);
        queue<int> q;

        for (const vector<int>& pre : prerequisites) {
            int u = pre[0];
            int v = pre[1];
            graph[u].push_back(v);
        }

        for (const vector<int>& u : graph) {
            for (int v : u) {
                in_degree[v]++;
            }
        }

        for (int i = 0; i < numCourses; i++) {
            if (in_degree[i] == 0)
                q.push(i);
        }

        vector<int> result;

        while (!q.empty()) {
            int u = q.front();
            q.pop();
            result.push_back(u);
            for (int v : graph[u]) {
                in_degree[v]--;
                if (in_degree[v] == 0)
                    q.push(v);
            }
        }

        return result.size() == numCourses;
    }
};
